### PR TITLE
feat: add etherscan client crate

### PR DESCRIPTION
**Pull request by mattsse**
## Motivation
Add etherscan.io client,
Ref #485 

## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [ ] Updated the changelog

### CODE REVIEW

File: ethers-etherscan/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+                    .unwrap(),
+            )
+            .await
+            .unwrap();
+    }
+
+    #[tokio::test]
+    #[ignore]
+    async fn can_verify_contract() {
+        // TODO this needs further investigation
```
Comment (gakonst) [inline ethers-etherscan/src/lib.rs:430]: Is this working now or does it still require more investigation?

File: ethers-etherscan/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+
+        // https://etherscan.io/address/0x9e744c9115b74834c0f33f4097f40c02a9ac5c33#code
+        let contract = include_str!("../resources/UniswapExchange.sol");
+        let address = "0x9e744c9115b74834c0f33f4097f40c02a9ac5c33"
+            .parse()
+            .unwrap();
+        let compiler_version = "v0.5.17+commit.d19bba13";
+        let constructor_args = "0x000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000005f5e1000000000000000000000000000000000000000000000000000000000000000007596179537761700000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000035941590000000000000000000000000000000000000000000000000000000000";
```
Comment (gakonst) [inline ethers-etherscan/src/lib.rs:438]: I wonder if we should move `BaseContract` to ethers-core now to encode constructor arguments in a nice way

File: ethers-etherscan/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+        self.optimization_used = Some("0".to_string());
+        self
+    }
+
+    pub fn code_format(mut self, code_format: CodeFormat) -> Self {
+        self.code_format = code_format;
+        self
+    }
+
+    pub fn evmversion(mut self, evmversion: impl Into<String>) -> Self {
```
Comment (gakonst) [inline ethers-etherscan/src/lib.rs:279]:
```suggestion
    pub fn evm_version(mut self, evm_version: impl Into<String>) -> Self {
```

File: ethers-etherscan/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+    optimization_used: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub runs: Option<String>,
+    /// NOTE: there is a typo in the etherscan API `constructorArguements`
+    #[serde(
+        rename = "constructorArguements",
+        skip_serializing_if = "Option::is_none"
+    )]
+    pub constructor_arguments: Option<String>,
+    pub evmversion: Option<String>,
```
Comment (gakonst) [inline ethers-etherscan/src/lib.rs:225]:
```suggestion
    pub evm_version: Option<String>,
```

File: ethers-etherscan/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+}
+
+/// The type that gets serialized as query
+#[derive(Debug, Serialize)]
+struct Query<'a, T: Serialize> {
+    apikey: Cow<'a, str>,
+    module: Cow<'a, str>,
+    action: Cow<'a, str>,
+    #[serde(flatten)]
+    other: T,
```
Comment (gakonst) [inline ethers-etherscan/src/lib.rs:198]: smart. never thought about that.

File: ethers-etherscan/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+    /// Get Contract Source Code for Verified Contract Source Codes
+    /// ```no_run
+    /// # use ethers_etherscan::Client;
+    ///
+    /// # #[tokio::main]
+    /// # async fn main() -> anyhow::Result<()> {
+    ///     let client = Client::new("mainnet", "API_KEY").unwrap();
+    ///     let meta = client
+    ///         .contract_source_code("0xBB9bc244D798123fDe783fCc1C72d3Bb8C189413".parse().unwrap())
+    ///         .await?;
```
Comment (gakonst) [inline ethers-etherscan/src/lib.rs:167]: Should we also compile this with Solc? Or provide some helper way to convert this to a `CompiledContract`

File: ethers-etherscan/src/lib.rs (hunk 1 above)
Comment (mattsse) [reply ethers-etherscan/src/lib.rs:430]: still get this response _instantly_, so I think I messed up the payload somehow, but tried various, alternatives like numbers as strings, etc.., will need to take a closer look at dapptools itself or any other etherscan client
```
    Response {
        status: "0",
        message: "NOTOK",
        result: "Error!",
    },
```

File: ethers-etherscan/src/lib.rs (hunk 2 above)
Comment (mattsse) [reply ethers-etherscan/src/lib.rs:438]: that'd be useful, I think we touched on abstracting functions as a trait via `abigen!` some time ago, I think it makes sense to follow up on this and move all general abstractions to core.

File: ethers-etherscan/src/lib.rs (hunk 6 above)
Comment (mattsse) [reply ethers-etherscan/src/lib.rs:167]: the etherscan api doesn't seem to provide the bytecode directly, however it is present in the html reponse with `div[id=verifiedbytecode2]` -.-
Additional tooling to instantly compile this is probably useful, I'll add this to the tracking issue

**Review by gakonst — APPROVED** [review]
Gg

**Pull request merged**